File: src/HashUtil.java
import java.io.*;
import java.util.*;

public class HashUtil {
	public static Pair<Integer, Integer> closest_repeated_pair(String[] list) {
		HashMap<String, Integer> map = new HashMap<String, Integer>();
		Pair<Integer, Integer> pair = null;
		for (int i = 0; i < list.length; i++) {
			if (map.containsKey(list[i])) {
				int j = map.get(list[i]);
				if (pair == null || pair.second - pair.first > i -j) {
					pair = new Pair(j, i);
				}
			}
			map.put(list[i], i);
		}
		return pair;
	}
	
	public static String anagram_hash(String s) {
		int[] count = new int[26];
		for (int i = 0; i < s.length(); i++) {
			count[s.charAt(i)-'a']++;
		}
		String hash = "";
		for (int i = 0; i < 26; i++) {
			hash += count[i];
			hash += (char)(i + 'a');
		}
		return hash;
	}
	
	public static Map<String, List<String>> get_anagram_sub_sets(List<String> list) {
		Map<String, List<String>> map =new HashMap<String, List<String>>();
		for (String s: list) {
			String key = anagram_hash(s);
			if (!map.containsKey(key)) {
				List<String> subset = new ArrayList<String>();
				subset.add(s);
				map.put(key, subset);
			} else {
				map.get(key).add(s);
			}
		}
		return map;
	}
	
	public static List<String> getDictionary(String filename) throws Exception{
		Scanner scanner = new Scanner(new File(filename));
		Set<String> set= new HashSet<String>();
		while (scanner.hasNext()) {
			String s = scanner.next();
			if (s.matches("[a-z]*"))
				set.add(s);
		}
		scanner.close();
		return new ArrayList(set);
	}
	
	public static Set<String> getDictionaryAsSet(String filename) throws Exception{
		Scanner scanner = new Scanner(new File(filename));
		Set<String> set= new HashSet<String>();
		while (scanner.hasNext()) {
			String s = scanner.next();
			if (s.matches("[a-z]*"))
				set.add(s);
		}
		scanner.close();
		return set;
	} 
	
	public static List<String> getFrequentItems(String input, int k) {
		Scanner scanner = new Scanner(input);
		HashMap<String, Integer> map = new HashMap<String, Integer>();
		int itemNum = 0;
		while (scanner.hasNext()) {			
			for (int i = 0; i <= k; i++) {
				if (scanner.hasNext()) {
					String s = scanner.next();
					itemNum++;
					if (map.containsKey(s)) {
						map.put(s, map.get(s) +1);
					} else if (map.size() <= k) {
						map.put(s, 1);
					}
				} else {
					break;
				}
			}
			if (map.size() == k +1) {
				Iterator it = map.entrySet().iterator();
				while (it.hasNext()) {
					Map.Entry<String, Integer> entry = (Map.Entry<String, Integer>)it.next();
					String s = entry.getKey();
					if (map.get(s) > 1) {
						map.put(s,  map.get(s) -1);
					} else {
						it.remove();
					}
				}
			}
		}
		scanner.close();
		scanner = new Scanner(input);
		for (String s: map.keySet()) {
			map.put(s, 0);
		}
		while (scanner.hasNext()) {
			String s = scanner.next();
			if (map.containsKey(s)) {
				map.put(s, map.get(s) + 1);
			}
		}
		List<String> list = new ArrayList<String>();
		for(String s: map.keySet()) {
			if (map.get(s) >= (double)itemNum/k) list.add(s);
		}
		return list;
	}
	
	public static Pair<Integer, Integer> find_smallest_subarray_covering_subset(String[] A, String[] Q) {
		HashSet<String> dict = new HashSet<String>(Arrays.asList(Q));
		HashMap<String, Integer> count_Q = new HashMap<String, Integer>();
		int l = 0, r = 0;
		Pair<Integer, Integer> res= new Pair(-1, -1);
		while (r < A.length) {
			// Keep moving r until it reaches end or count_Q has |Q| items
			while (r < A.length && count_Q.size() < Q.length) {
				if (dict.contains(A[r])) {
					if (count_Q.containsKey(A[r])) {
						count_Q.put(A[r], count_Q.get(A[r])+1);
					} else {
						count_Q.put(A[r], 1);
					}					
				}
				++r;
			}
			if (count_Q.size() == Q.length && (res.first == -1 && res.second == -1)) {
				res.first = l;
				res.second = r -1;
			}
			// Keep moving l until it reaches end or count_Q has less than |Q| items
			while (l < A.length && count_Q.size() == Q.length) {
				if (dict.contains(A[l])) {
					if (count_Q.get(A[l]) == 1) {						
						count_Q.remove(A[l]);
						if ((res.first == -1 && res.second == -1) || r-1-l < res.second - res.first) {
							res.first = l;
							res.second = r -1;
						}
					} else {
						count_Q.put(A[l],count_Q.get(A[l]) -1);
					}
				}
				++l;
			}			
		}
		return res;
	}
	
	public static Pair<Integer, Integer> find_smallest_sequentially_covering_subset(String[] A, String[] Q) {
		HashMap<String, Integer> K = new HashMap<String, Integer>();
		int[] L = new int[Q.length];
		Arrays.fill(L, -1);
		int[] D = new int[Q.length];
		Arrays.fill(D, Integer.MAX_VALUE);
		// Initialize K
		for (int i = 0; i < Q.length; ++i) {
			K.put(Q[i], i);
		}
		Pair<Integer, Integer> res = new Pair(-1, A.length); // default value
		for (int i = 0; i < A.length; ++i) {
			if (K.containsKey(A[i])) {
				int id = K.get(A[i]);
				if (id == 0) { // first one, no predecessor
					D[0] = 1;
				} else if(D[id -1] != Integer.MAX_VALUE) {
					D[id] = i - L[id-1] + D[id-1];
				}
				L[id] = i;
				if (id == Q.length -1 && D[id] < res.second - res.first +1) {
					res.first = i - D[id] + 1;
					res.second = i;
				}
			}
		}
		return res;
	}

}

File: C++/HashUtil.cpp
#include <unordered_map>
#include <unordered_set>
#include <limits>
#include <iostream>
#include <algorithm>
#include <iterator>
#include <sstream>
#include <list>
#include "HashUtil.h"

using namespace std;

int find_closest_repetition(const vector<string> &s) {
    unordered_map<string, int> string_to_location;
    int closest_distance = numeric_limits<int>::max();
    for ( int i = 0; i < (int)s.size(); i++) {
        auto it = string_to_location.find(s[i]);
        if (it != string_to_location.end()) {
            closest_distance = min(closest_distance, i - it->second);
        }
        string_to_location[s[i]] = i;
    }
    return closest_distance;
}

void find_anagrams(const vector<string> &dictionary) {
    unordered_map<string, vector<string>> hash;
    for (const string &s: dictionary){
        string sorted_str(s);
        sort(sorted_str.begin(), sorted_str.end());
        hash[sorted_str].emplace_back(s);
    }
    for (const pair<string, vector<string>> &p : hash) {
        if (p.second.size() >=2) {
            copy(p.second.begin(), p.second.end(), ostream_iterator<string>(cout, " "));
            cout << endl;
        }
    }
}

vector<string> search_frequent_items(istringstream &sin, const int &k) {
    string buf;
    unordered_map<string, int> hash;
    int n = 0;
    while (sin >> buf) {
        ++hash[buf], ++n;
        // Detecting k+1 items in hash, at least one of them must have exactly 1 in it.
        // We will discard those k + 1 items by 1 for each
        if (hash.size() == (unsigned int)(k + 1)) {
            auto it = hash.begin();
            while (it != hash.end()) {
                if (--(it->second) == 0) {
                    hash.erase(it++);
                } else {
                    ++it;
                }
            }
        }
    }
    
    // Reset hash for the following counting
    for (auto &it: hash) {
        it.second = 0;
    }
    // Reset the stream and read it again
    sin.clear();
    sin.seekg(0, ios::beg);
    // Count the occurence of each candidate word
    while (sin >> buf) {
        auto it = hash.find(buf);
        if (it != hash.end()) {
            ++it->second;
        }
    }
    vector<string> ret;
    for (const pair<string, int> &it : hash) {
        // Select the word which occurs >= n/k times
        if (it.second >= static_cast<double>(n)/k) {
            ret.emplace_back(it.first);
        }
    }
    return ret;
}

pair<int, int> find_smallest_subarray_covering_subset(const vector<string> &A, const vector<string> &Q) {
    unordered_set<string> dict(Q.cbegin(), Q.cend());
    unordered_map<string, int> count_Q;
    int l =0, r = 0;
    pair<int, int> res(-1, -1);
    while (r < static_cast<int> (A.size())) {
        // Keep moving r until it reaches end or count_Q has |Q| items
        while (r < static_cast<int> (A.size()) && count_Q.size() < Q.size()) {
            if (dict.find(A[r]) != dict.end()) {
                ++count_Q[A[r]];
            }
            ++r;
        }
        if (count_Q.size() == Q.size() && ((res.first == -1 && res.second == -1) || r -1 -l < res.second - res.first)) {
            res = {l, r -1};
        }
        // Keep moving l until it reaches end or count_Q has less than |Q| items
        while (l < static_cast<int>(A.size()) && count_Q.size() == Q.size()) {
            if (dict.find(A[l]) != dict.end()) {
                auto it = count_Q.find(A[l]);
                if (--(it->second) == 0) {
                    count_Q.erase(it);
                    if ((res.first == -1 && res.second == -1) || r -1 -l < res.second - res.first) {
                        res = {l, r-1};
                    }
                }
            }
            ++l;
        }
    }
    return res;
}

pair<int, int> improved_find_smallest_subarray_covering_subset(istringstream &sin, const vector<string> &Q) {
    list<int> loc; // Tracking the last occurrence (index) of each string in Q
    unordered_map <string, list<int>::iterator> dict;
    for (const string &s: Q) {
        dict.emplace(s, loc.end());        
    }
    pair<int, int> res(-1, -1);
    int idx = 0;
    string s;
    while (sin >> s) {
        auto it = dict.find(s);
        if (it != dict.end()) { // s is in Q
            if (it->second != loc.end()) {
                loc.erase(it->second);
            }
            loc.emplace_back(idx);
            it->second = --loc.end();
        }
        if (loc.size() == Q.size() && // find |Q| keywords
            ((res.first == -1 && res.second == -1) || idx - loc.front() < res.second - res.first) ) {
                res = {loc.front(), idx};
        }
        ++idx;
    }
    return res;
}

pair<int, int> find_smallest_sequentially_covering_subset(
    const vector<string> &A, const vector<string> &Q) {
    unordered_map <string, int> K;
    vector<int> L(Q.size(), -1), D(Q.size(), numeric_limits<int>::max());
    // Initialize K
    for (unsigned int i = 0; i < Q.size(); ++i) {
        K.emplace(Q[i], i);        
    }
    pair<int, int> res(-1, A.size()); // default value
    for (unsigned int i = 0; i < A.size(); ++i) {
        auto it = K.find(A[i]);
        if (it != K.cend()) {
            if (it->second == 0) { // first one, no predecessor
                D[0] = 1; // base condition
            } else if (D[it->second -1] != numeric_limits<int>::max()) {
                D[it->second] = i - L[it->second -1] + D[it->second -1];                
            }
            L[it->second] = i;
            if (it->second == Q.size() - 1 && D.back() < res.second - res.first + 1) {
                res = { i - D.back() + 1, i};
            }
        }
    }
    return res;        
}